velox/common/memory/SharedArbitrator.cpp

```diff
@@ -237,19 +166,135 @@ SharedArbitrator::~SharedArbitrator() {
   }
 }
 
-void SharedArbitrator::getCandidateStats(
+void SharedArbitrator::addPool(const std::shared_ptr<MemoryPool>& pool) {
+  VELOX_CHECK_EQ(pool->capacity(), 0);
+  {
+    std::unique_lock guard{poolLock_};
+    VELOX_CHECK_EQ(candidates_.count(pool.get()), 0);
+    candidates_.emplace(pool.get(), pool);
+  }
+
+  std::lock_guard<std::mutex> l(stateLock_);
+  const int64_t maxBytesToReserve =
+      std::min(maxGrowCapacity(*pool), memoryPoolInitialCapacity_);
+  const int64_t minBytesToReserve = minGrowCapacity(*pool);
+  const uint64_t reservedBytes =
+      decrementFreeCapacityLocked(maxBytesToReserve, minBytesToReserve);
+  try {
+    checkedGrow(pool.get(), reservedBytes, 0);
+  } catch (const VeloxRuntimeError&) {
+    incrementFreeCapacityLocked(reservedBytes);
+  }
+}
+
+void SharedArbitrator::removePool(MemoryPool* pool) {
+  VELOX_CHECK_EQ(pool->reservedBytes(), 0);
+  shrinkCapacity(pool, pool->capacity());
+
+  std::unique_lock guard{poolLock_};
+  const auto ret = candidates_.erase(pool);
+  VELOX_CHECK_EQ(ret, 1);
+}
+
+void SharedArbitrator::getCandidates(
     ArbitrationOperation* op,
     bool freeCapacityOnly) {
   op->candidates.clear();
-  op->candidates.reserve(op->candidatePools.size());
-  for (const auto& pool : op->candidatePools) {
-    const bool selfCandidate = op->requestRoot == pool.get();
+
+  std::shared_lock guard{poolLock_};
+  op->candidates.reserve(candidates_.size());
+  for (const auto& candidate : candidates_) {
+    const bool selfCandidate = op->requestRoot == candidate.first;
+    std::shared_ptr<MemoryPool> pool = candidate.second.lock();
+    if (pool == nullptr) {
+      VELOX_CHECK(!selfCandidate);
```

Comment:
Maybe we can have a check flag 'selfPoolDestroyed' outside and move the check at the bottom outside of the for loop to save some cycles

Reply:
Not sure if I get your question. This is to check unexpected self destroyed pool and the check should be pretty lightweight.